### ISSUE TITLE
Questions about pre-computed particles simulation 

### ISSUE THREAD

**Issue by Rolandxx7** [astra-vision/rain-rendering#23]
Dear author @rdecharette ,

Thanks for your great contribution!

I would like to ask 4 import questions:
  1. Why can't your pre-computed particles simulation be used for my own dataset? For example, why can't KITTI's pre-computed particles simulation be used in Cityscapes dataset? For settings["cam_pos"] = [1.5, 1.5, 0.3], where is the origin of its coordinates?
  2. As your pre-computed particles simulation of KITTI dataset is for image_2, if I want to render image_3 of KITTI dataset, can I reuse this pre-computed particles simulation?
  3. If I want to re-generate a different pre-computed particles simulation for KITTI, is it possible? What is it determined by?
  4. For an one frame, is the position of the generated rain streak by [weather particle simulator](https://github.com/cv-rits/weather-particle-simulator) random? Can we control the location of the rain streak generated from [weather particle simulator](https://github.com/cv-rits/weather-particle-simulator).

**Comment by rdecharette** [comment]
Hello @Rolandxx7 

Thanks ;)

1. Particle simulation computes the position of particles in the image space which is dependent on camera extrinsics and intrinsics, so it's unlikely that your own dataset has the exact same optics, lens, cam pos/orientation. That said, technically, there is no limitation you can completely use the KITTI particles simulation on your dataset it just won't be really realistic. If your setup is sufficiently close to KITTI it won't be noticed.
2. You're question got me thinking. But you are right, since cam2 / cam3 are same height and sensors (cf. [setup](https://www.cvlibs.net/datasets/kitti/setup.php)) you can use our simulation for both.
3. Not 100% sure to understand the question. Short answer: no. Long answer: To avoid artefacts, particles are initialized randomly outside (above) the camera FOV at their raindrops terminal velocity, as a function of rainfall rate and location model (the default model is Montreal-located it includes pressure, air friction, Drop Size Distribution, etc.. Others locations are available in the simulator). Initializing rainstreak is technically feasible in the code (though code is not open) but more importantly if manual it won't respect the physics.

Hope that helps

**Comment by Rolandxx7** [comment]
> 
Dear author @rdecharette ,

Thank you for such a detailed explanation!
I would like to ask 4 more questions.

1. For the config file named kitti.py, I would like to know what the settings["cam_focus_plane"] means. Also, what is the difference between "cam_focus_plane" and "cam_focal"? I can't figure out why "cam_focus_plane" is in meters, as the focal point is in the middle of the optical center and the imaging plane.
https://github.com/astra-vision/rain-rendering/blob/master/config/kitti.py#L60
2. For settings["cam_pos"] = [1.5, 1.5, 0.3] in kitti.py, where is the origin of its coordinates?
https://github.com/astra-vision/rain-rendering/blob/master/config/kitti.py#L64
3. If I re-generate the rain streak with the [weather particle simulator](https://github.com/cv-rits/weather-particle-simulator) for the KITTI dataset, will the result be the same as your pre-computed particles simulation of KITTI dataset?
4. I want most of the raindrops to appear in my specified image area instead of randomly. How should I achieve this?

**Comment by rdecharette** [comment]
Dear @Rolandxx7

1. The focus plane is the distance that appears net on the imaging device. Below or beyond that distance, rainstreaks are slightly blurred with the circle of confusion computed here: https://github.com/astra-vision/rain-rendering/blob/9e994b0cd7e5e210189821345f6655b21c79ca48/common/bad_weather.py#L464 However, I just discovered that the value is overwritten to 6.0 meters anyway here: https://github.com/astra-vision/rain-rendering/blob/9e994b0cd7e5e210189821345f6655b21c79ca48/common/generator.py#L267
Anyhow this has minor visual impact.
2. Asfair (it's quite far away ;)) it's not crucial to change this because the particle volume is automatically readjust to cover the camera FOV. The important coordinate is y=1.5m which is altitude (as drops disappear when hitting virtual ground = 0 altitude). In my memory the rationale was to set origin at the front of the car at ground-level, but again it's not crucial.
3. Yes, though different platform rounding may deviate slightly. I remember running unit test showing reproducible results.
4. First, this is not really designed for that, since if you filter out some drops then the rainfall rate will **not** be realistic. That said, you have two options. 

- (easier) Changing rendering you could filter some drops when loading the file ([here](https://github.com/astra-vision/rain-rendering/blob/9e994b0cd7e5e210189821345f6655b21c79ca48/common/bad_weather.py#L200-L239)), or skip them when rendered ([here](https://github.com/astra-vision/rain-rendering/blob/9e994b0cd7e5e210189821345f6655b21c79ca48/common/generator.py#L435)). 
- (harder) Changing particle simulator you could adjust the particle emitter volume (for ex so you have raindrop falling only on the left part of the image) which is option 411 (for rain) in the simulator menu. Here you'll keep accurate rainfall rate but you need to re-compute particles and render.

**Comment by Rolandxx7** [comment]
Dear author @rdecharette ,

Thank you so much! Your quick reply is so helpful!

I'm sorry, I still have 3 questions that I don't understand.
1. I would like like to distribute the rain streak around the image coordinates of (120,200), (500,20), (1000,300), while keeping the rainfall constant, what should I do?  Is this possible by changing the particle simulator, but you said before that your particle simulator is not open source.
2. For the 32921668_2.0ms_10fps_camera0.xml
`<r pid="1100331" wp1="[-0.274533;-0.104401;-0.949732]" wd1="0.00100009" wp2="[-0.274533;-0.112396;-0.894176]" wd2="0.00100009" ip1="[248.014;45.659]" iw1="1.35874" ip2="[224.84;23.866]" iw2="1.44316" />`
Why are there only two coordinates, wp1 and wp2?
3. In the code, except for the xxx_2.0ms_10fps_camera0.xml file, the rest of the .xml file(for ex xxx_2.0ms_10fps_projector0.xml) does not seem to be used, right? Also, how should I be able to know the meaning of several other .xml files?

**Comment by rdecharette** [comment]
1. Let me clarify. In the first message you asked how to initialize a streak manually => this is not possible without the simulator code. But in second message you asked how to get rain only on a part of the image which is different and feasible just by modifying the particle emitter pos: option 411 (rain) in the simulator menu. Binaries for the simulator are available [here](https://github.com/astra-vision/weather-particle-simulator)
2. Particle simulator defines streaks as a line, so two points ((x1,y1,z1), (x2,y2,z2)) in the 3D world. wd1/wd2 are the starting and end radius. Then rain-rendering approximates the streak photometry.
3. True. In this project we care only about the position of drops in the camera space. But initially the particle simulator was done for a completely different project ([here](http://www.cs.cmu.edu/smartheadlight/index2.html)), hence the other files which allows deeper analysis.

Hope that is clearer for you.

**Comment by Rolandxx7** [comment]
Dear author @rdecharette,

Happy Weekend! Thanks for your detailed reply!

For the `32921668_2.0ms_10fps_camera0.xml` file, I don't understand why there are only 350 raindrops for one frame of the image in the 3D world. If it were the real 3D world, the number of raindrops would be much more than 350 in the FOV view. Could you explain this to me?

**Comment by rdecharette** [comment]
Hello again,

This is described in Sec. 3.1.1 of our paper ([here](https://arxiv.org/pdf/2009.03683.pdf)): rendering all drops would make no sense, and we only render drops that project on at least 1px of the imaging device. Any drops that project on a smaller area, is considered to be "fog-like rain" (as originally defined by Garg and Nayar, Fig. 3 [here](https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=1544839)) and simulated as a light volume attenuation.

**Comment by Rolandxx7** [comment]
Hi @rdecharette,

However, I think even if we ignore the raindrops that are imaged at less than 1px, the number of remaining raindrops in the 3D world should be much more than 350. So could you explain why the number of drops is only around 350?

**Comment by rdecharette** [comment]
As mentioned, these are just >1px drop so I don't see why 350 is surprising ?
Also not sure of which rain fallrate you're mentioning. Checking a random KITTI data_object file I get: 118/350/1348 drops for 10/25/100 mm/hr rain. This seems reasonable. You can check [our adaptive headlight paper](http://www.cs.cmu.edu/~ILIM/publications/PDFs/DTBRKN-ICCP12.pdf) for deeper insight.

**Comment by Rolandxx7** [comment]
Dear author @rdecharette,

Sorry to open this issue again!

I would like to know why I cannot reproduce the effect of the rain-augmented datasets you published for the KITTI dataset with 25mm/hr rainfall. For example, for the rain_mask of 000000.png, I found that my regenerated image is not the same as the image in your rain-augmented dataset. Could you tell me why?

**Comment by rdecharette** [comment]
Hello, can you detail how you regenerated this rain mask? Which command? Did you use the python wrapper or did you run the particle simulator directly? 
Keep in mind that different platforms have different rounding hence different simulations.

**Comment by Rolandxx7** [comment]
Hi @rdecharette,
Thank you for responding to my question!

I regenerated this rain mask exactly according to the KITTI section of your [README.md](https://github.com/astra-vision/rain-rendering#readme) (using pre-computed KITTI particles simulations) and use this command `python main.py --dataset kitti --intensity 25 --frame_end 10`. So I'm wondering if your published KITTI rain-augmented dataset wasn't generated according to the [README.md](https://github.com/astra-vision/rain-rendering#readme) .

Besides, I have two below questions about the `wp1=[-0.274533; -0.104401; -0.949732]` in the 32921668_2.0ms_10fps_camera0.xml:
1. Are the coordinates of wp1 under the camera coordinate system of KITTI dataset? If not, what is it？What is the conversion relationship between this coordinate system and the camera coordinate system？
2. For the coordinates of wp1, what is the orientation of x-axis, y-axis and z-axis respectively？

**Comment by rdecharette** [comment]
Hello @Rolandxx7 , 

That is intriguing indeed. I double checked the guidelines and got the exact same image as you.
![Screenshot from 2023-04-12 11-45-00](https://user-images.githubusercontent.com/5967026/231420714-e14e1734-0f58-4f35-b6de-c0d8f8ed5161.png)
Upon closer inspection, the renderer is correctly rendering the particles listed in simulation files. Hence, I suppose the shared dataset use particles initialized with a different seed than particles files. I suspect this happened in 2020 as we had to recompute a number of simulation due to files loss. If they are statistically similar, the difference on downstream task should be negligible.

1,2. I do not recall but looking at the code: the coordinate system is the same as KITTI in the renderer (x: pointing right, z: pointing front, y: point upward). 
However, I recall that in particles simulator Y is instead pointing downward, following physic convention, hence the reversal of image and world space upon particle loading in the renderer ([here](https://github.com/astra-vision/rain-rendering/blob/9e994b0cd7e5e210189821345f6655b21c79ca48/common/bad_weather.py#L221-L224)).
3. About wp1, wp2, wd1 are in meters.

**Comment by rdecharette** [comment]
This is the red coordinate system in the **renderer** (again, not in particle simulator). I indeed had a typo in my previous answer (fixed). I hope it's correct now, but that said you could also simply debug Python to find out the correct coordinate system (for ex, by rendering drops one by one and looking at their world vs image pos).

As for negative / positive coordinates, you probably made a mistake. There are approx half of the particles with positive X or Y coordinates, as expected. A simple regex allows finding particles with positive X coordinate: `wp1="\[[0-9\.]+;-?[0-9\.]+`, same for positive Y with: `wp1="\[-?[0-9\.]+;[0-9\.]+`.

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

There is still a question for negative / positive coordinates.
Taking `wp1=[-0.274533; -0.104401; -0.949732] ` as an example, what I don't understand is why the z coordinate of some drops is less than 0. As we know, if z is less than 0, which means that the drop is directly in behind the camera. However, the camera can't record the raindrops behind it.

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

I have tried to render drops one by one and look at their world vs image pos. 
However, I find there is a problem. 

For example, if you want to render the 000000.png of KITTI dataset, you should use the calibration file corresponding to the 000000.png to know the image position of the drops. Moreover, the calibration files vary from each image in the KITTI dataset, that is to say, the camera intrinsics P2 of each image are different. 

As I said, the camera intrinsics are different for each picture in KITTI, so I wonder why you can know the image coordinates corresponding to the raindrops in the `32921668_2.0ms_10fps_camera0.xml` file before knowing which image to render. (for ex, `wp1="[-0.274533;-0.104401;-0.949732]", ip1="[248.014;45.659]"`)

**Comment by rdecharette** [comment]
you're right. In fact frame-wise calibration is neglected.

In the physical simulator, we use a pinhole model so here it uses KITTI camera intrinsics+extrinsic, ie. camera sensor+lens characteristics. Hence, we ignore radial distortion of the lens -- otherwise a streak could **not** be approximated as a start and end point in the image space, it should be a set of points which would drastically slow down rendering.
In practice I think this makes very little difference. Probably a small portion of streaks are shifted by a few pixels. But if you want to recompute the precise image projection, you could do so with their world coordinates: wp1,wp2,wd1,wd2.

In the renderer, asfair latest version ignores calibs too because. I remember having used them to reweight the envmap FOV of each drop, which led to uperceivable realism improvement at a high computational cost.

Best.

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

Thanks for your detailed explanation.

I would like to know how I can know the KITTI camera intrinsics+ extrinsics you used. Because I only have the camera intrinsics corresponding to each image in KITTI.

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

I've read this KITTI [journal paper](https://www.cvlibs.net/publications/Geiger2013IJRR.pdf) and didn't find the manufacturer spreadsheet for KITTI camera/lens.

Also, I would like to know what is Camera 0 IC matrix? what is it used for?

**Comment by rdecharette** [comment]
Hi @Rolandxx7,

please have more detailed look. All infos are already in my previous reply. Section II of the journal paper details sensors/lenses, then googling provides manufacturer sheets.
Yes image has intrinsics written and extrinsics in lines 11-15.
IC is the standard abbreviation for Initial Condition. Since camera has motion, line 15 is the IC camera matrix at the beginning of the simulation (before cam moves).

**Comment by rdecharette** [comment]
Camera IC matrix matters for simulation, it defines the initial state of the simulation. Yes it is Camera 0 (the only cam here), of course in world coordinate system.

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

Sorry for so many questions.

I still have 4 questions that I don't understand.
1. Why do you need to record the initial state of the simulation of Camera 0? 
2. Don't we just project the raindrop of the camera coordinate system onto the image, this has nothing to do with the world coordinate system?
3. I want to make sure again, so the raindrops `wp1` are in the camera coordinate system, and the Camera IC matrix uses the world coordinate system? The world coordinate system is the green in the picture below? 
![image](https://user-images.githubusercontent.com/76581571/232182673-dffba2bc-518e-42f2-a8eb-2673155be648.png)
4. why do I see that the diameter of droplets in some papers is 0.05mm, which is far from the one(0.001m) in your paper.

**Comment by rdecharette** [comment]
Hello @Rolandxx7,
I'm sorry but I cannot guide you through all the details of the paper as this is already 24 messages long.
8h

**Comment by Rolandxx7** [comment]
Hello @rdecharette,

I am sure these will be my last 4 questions. 
Could you please help me for the last time, as I can not find the answer from your paper.

Sincerely.

**Comment by rdecharette** [comment]
Hello @Rolandxx7,

I'm sorry but indeed I cannot guide you through the whole paper step by step. Especially, several questions in the thread find answers by reading the literature and playing with the simulator.

1. Not sure to understand. Initial conditions are required for any simulator to set state at t=0. We record it just to log all info.
2. I admit our naming might be confusing here. Let me rephrase: we consider (A) the 3D global world coordinate system fixed at (0,0,0) and (B) the 3D camera world coordinate system which pos is initialized with IC and moves according to the (car) motion. Physics simulation happens in (A), then each particle coordinates is converted to (B) to compute its image projection.
3. Our global world coordinate system (A) of the simulator has nothing to do with KITTI. `*_Camera0.xml` contains the particles `wp1`, `wp2` positions in the "3D camera world coordinate system" (B). Instead `*_particles.xml` shows the `z` of particles as they hit the ground in global world coordinate (A). For example for 25mm/hr rain, in `32921668_2.0ms_10fps_camera0.xml` particle ID 2764456 has z ~= -5.15 (B) and z=-284 (A) in `32921668_2.0ms_10fps_particles.xml` -- which is logical since camera moved in negative z direction. 
4. Droplets is not a well defined notion in the literature as there is no physical boundaries between doplets/drop. We follow common practice and use 1mm but other may use 0.5mm, 2mm, etc. This can be changed in the simulator (menu 412) -- but would imply updating fog-like rain simulation too. 

Our simulator / renderer is largely re-used, so I believe with some effort the remaining doubts should solve.
